Ipa interface overview (#1)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/IPAuthoring_interfaces_overview/Readme.md

Reviewer: don't forget to add a `sample.json`

IP Authoring Interface overview

> ==================

Reviewer: ```suggestion
# IP Authoring Interfaces Overview
```

> One of the most powerful features of IP Authoring is that you can control how your IP interacts with the rest of your design. When building a full-stack system (e.g. using buffers/accessors) the oneAPI compiler will infer a set of interfaces for you that fits with your selected BSP. The new IP Authoring features let you customize the interfaces for your standalone IP, without requiring corresponding attach points in a BSP.

Reviewer: ```suggestion
One of the most powerful features of IP Authoring is that you can configure how your FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler interacts with the rest of your custom FPGA design. When targeting the full-stack flow (e.g. using buffers/accessors) the Intel® oneAPI DDPC++/C++ Compiler will infer a set of interfaces for you that fits with your selected BSP. When you create FPGA IP with the Intel® oneAPI DPC++/C++ Compiler, you can customize the interfaces for your IP component, since IP components are not constrained by specific attach points in a specific BSP.
```

| Area                 | Description
|:---                  |:---

> | What you will learn  | How to choose between different invocation and date interface for IP Authoring, and how to implement and config them

Reviewer: ```suggestion
| What you will learn  | The basics of different invocation interfaces and data interfaces for FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler, and how to implement and configure these interfaces
```

| Time to complete     | 60 minutes
| Category             | Getting Started

## Prerequisites

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 18.04/20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10
| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software             | Intel® oneAPI icpx Compiler

> > **Note**: Even though the Intel icpx OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.

Reviewer: ```suggestion
> **Note**: Even though the Intel® oneAPI DDPC++/C++ Compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
```

double-check the product name. note: it's *oneAPI* not *OneAPI*. I won't mark all of them

   tier1("Tier 1: Get Started")
   tier2("Tier 2: Explore the Fundamentals")
   tier3("Tier 3: Explore the Advanced Techniques")
   tier4("Tier 4: Explore the Reference Designs")

   tier1 --> tier2 --> tier3 --> tier4

   style tier1 fill:#f96,stroke:#333,stroke-width:1px,color:#fff
   style tier2 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier3 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
   style tier4 fill:#0071c1,stroke:#0071c1,stroke-width:1px,color:#fff
```

Find more information about how to navigate this part of the code samples in the [FPGA top-level README.md](/DirectProgramming/C++SYCL_FPGA/README.md).
You can also find more information about [troubleshooting build errors](/DirectProgramming/C++SYCL_FPGA/README.md#troubleshooting), [running the sample on the Intel® DevCloud](/DirectProgramming/C++SYCL_FPGA/README.md#build-and-run-the-samples-on-intel-devcloud-optional), [using Visual Studio Code with the code samples](/DirectProgramming/C++SYCL_FPGA/README.md#use-visual-studio-code-vs-code-optional), [links to selected documentation](/DirectProgramming/C++SYCL_FPGA/README.md#documentation), and more.

## Invocation Interface

> The Intel icpx compiler generates an interface for your SYCL* kernel. The invocation interface controls how the component communicates with the rest of your system. This interface lets your wider system start the component, and lets the component inform the system when it is finished.

Reviewer: ```suggestion
The invocation interface is generated by the Intel® oneAPI DDPC++/C++ Compiler and controls how your FPGA IP communicates with the rest of your system. This interface lets your wider system start the component, and lets the component inform the system when it is finished. There are two ways that the invocation interface can be implemented: 

1. As signals mapped to the IP's control/status register (CSR)

2. As a handshake similar to the ready/valid handshake used by streaming interfaces.
```

> ### MM Agent

Reviewer: ```suggestion
### Register-mapped
```
descriptions should match the property name so the property name doesn't feel out-of-place

> The signals for starting your component, and the signals used by the component to signal its status are stored in a MM Agent space. You should use an MM Agent invocation interface if your component will sit in a system like this: \

Reviewer: ```suggestion
The signals for starting your component, and the signals used by the component to signal its status are mapped to registers accessible through an Avalon memory-mapped agent interface. You should use a register-mapped invocation interface if your component will operate in a system like this:
```

> ![](Readme.resource/invocation_mm_agent.png)

Reviewer: ```suggestion

![](Readme.resource/invocation_mm_agent.png)
```

leave an empty line in markdown to force a newline

> By default, the Intel icpx compiler generate MM Agent invocation interface for kernels, as you may observed in the [Naive Design](naive).

Reviewer: ```suggestion
By default, the Intel® oneAPI DDPC++/C++ Compiler generates a register-mapped invocation interface for kernels, as you may observe in the [Naive Design](naive).
```

### Streaming

> The signals for starting your component, and the signals used by the component to signal its status are stored in a MM Agent space. You should use an MM Agent invocation interface if your component will sit in a system like this: \

Reviewer 2: I wonder if this is correct? It is identical to MM Agent description above and mentioning  "You should use an **_MM Agent_** invocation ..."

Reviewer: I think you copy-pasted but didn't read what you copy-pasted :) the [source material](https://wiki.ith.intel.com/pages/viewpage.action?pageId=1718680879#MigratinganHLSdesigntoDPC++-ComponentInvocationInterface) has a mistake.

Author: you are correct, this is a mistake

> The signals for starting your component, and the signals used by the component to signal its status are stored in a MM Agent space. You should use an MM Agent invocation interface if your component will sit in a system like this: \

Reviewer: ```suggestion
The signals for starting your component and the signals used by the component to signal its status appear as dedicated conduits on your IP component. These signals mimic the ready-valid handshake commonly used by streaming interfaces. You should use a streaming invocation interface if your component will operate in a system like this:
```

![](Readme.resource/invocation_streaming.png)

For detail on how to invoke a streaming invocation interface, go to [handshake-control](handshake-control)

## Data Interface

> The data interfaces control how data enters and leaves your component's data path. Interface customization is not yet part of the official oneAPI release, so you need to include a 'prototype' header files in order to use this feature. When you compile, you need to include the flag `-I$INTELFPGAOCLSDKROOT/include` to include the interfaces header files.

Reviewer: ```suggestion
The data interfaces allow data to enter and leave your component's data path. There are three kinds of data interface:

1. Memory-mapped host interface

2. Streaming interface

3. Memory-mapped agent (CSR)
```

### Memory-Mapped Host

> When your data resides in an off-kernel memory (e.g. global memory or DDR, or even an on-chip block RAM) you can use an MM Host data interface to communicate with it. MM Host interfaces are inferred by default for all pointer members in oneAPI kernels, as you may observed in the [Naive Design](naive). \

Reviewer: ```suggestion
When your data resides in an off-kernel memory (for example: global memory or off-chip DRAM, or an on-chip block RAM) you can use a memory-mapped host data interface to communicate with it. Memory-mapped host  interfaces are inferred by default for all pointer members in oneAPI kernels, as you may observe in the [Naive Design](naive).
```

![](Readme.resource/data_mm_host.png)

> You can customize this interface by explicitly declare interfaces with type `annotated_arg` or `annotated_ptr`. See [mm-host](mm-host/) for example of how to do this.

Reviewer: ```suggestion
You can customize this interface by explicitly declare interfaces with type `annotated_arg` or `annotated_ptr`. See [mm-host](mm-host/) for an example of how to do this.
```

TODO: add a link to Hakim's MM Host tutorial once it's published for more detail https://github.com/HakimTem/oneAPI-samples/pull/1

### Stream (Pipe)

> You may also choose to stream your input or output data into/out of your kernel using pipes. Pipes are a first-in first-out (FIFO) buffer construct that provide links between

Reviewer: ```suggestion
You may also choose to stream data into/out of your kernel using pipes. Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. The Intel® oneAPI DPC++/C++ Compiler provides host pipes for host-kernel streaming and inter-kernel pipes for kernel-kernel streaming. The [host-pipes](host-pipes/) example demonstrates how to use host pipes to read and write data through an Avalon streaming interface that can be connected to other IP components in an FPGA design. 
```

elements of a design. The Intel® oneAPI DPC++/C++ Compiler provides Host Pipe for host-kernel streaming and Cross-kernel Pipe for kernel-kernel streaming. \
![](Readme.resource/data_streaming.png)

> For a kernel that is designed to interface with the outside world, you can use host pipes. For example of using Host Pipe, goto [host-pipes](host-pipes/)

Reviewer: ```suggestion
```

> ### Memory-Mapped Agent

Reviewer: ```suggestion
### Memory-Mapped Agent (Control/Status Register)
```

> You may wish to register your data in a memory-mapped register space. \

Reviewer: ```suggestion
You may wish to register your data in a memory-mapped register space.

```

use an empty line to get a newline. i won't correct this anymore

![](Readme.resource/data_mm_agent.png)

There are two ways you can do this:
- Use (non-pointer)members of your functor. This is the default way to get data into your kernel as demonstrated in the [Naive Design](naive).

> - Use Host Pipes configured with the `sycl::ext::intel::experimental::protocol_avalon_mm` as protocol.

Reviewer: ```suggestion
- Use host pipes configured to use the `sycl::ext::intel::experimental::protocol_avalon_mm` protocol.
```

'Host Pipes' is not a proper noun, so it shouldn't be capitalized. I won't correct this anymore, but please fix it elsewhere in the readmes

> Unfortunately, oneAPI kernels do not have return types, so you must use a Host Pipe to write the output. If you would like to store your output in the CSR for an MM Host device to access (as shown in the graphic above), please review the [csr-pipes](csr-pipes/) example design.

Reviewer: ```suggestion
oneAPI kernels do not have return types, so you must use one of the three data interfaces to pass output data. If you would like to store your output in the CSR for a memory-mapped host device to access (as shown in the graphic above), please review the [csr-pipes](csr-pipes/) example design.
```

## Build the Design

> There are 5 different example designs in this sample. These build targets will work for all designs. Use these commands to run each design, depending on your OS.

Reviewer: ```suggestion
There are 5 different example designs in this sample, all of which implement a simple vector addition. You can compare the C++ source code for each of these designs.

1. [Naive](naive/) This simple vector add implementation is similar to the one presented in the [fpga_compile](LINK PLEASE) code sample. It demonstrates a kernel that uses a register-mapped invocation interface, and a single shared memory-mapped host data interface. This is the only design that can support the full-system compilation flow, since the other designs implement interface customizations that are not supported by any BSP.

2. [Streaming invocation](handshake-control/) This implementation uses a streaming invocation interface and conduits to pass the input pointer arguments instead of the CSR. It uses a single, shared memory-mapped host data interface.

3. [Customized memory-mapped host data interface](mm-host/) This implementation uses a register-mapped invocation interface, and demonstrates how to use `annotated_ptr` to customize the memory-mapped host data interface.

4. [Streaming Data (host pipes)](host-pipes/) This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to configure a streaming data interface.

5. [CSR Data](csr-pipes/) This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to configure a streaming data interface, and how to return output into the CSR using the `protocol_avalon_mm` pipe protocol.

```

> All designs supports four compilation options: Emulator, Simulator, Optimization Report, FPGA Hardware. For details on the different compilation options, see [fpga_compile](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/GettingStarted/fpga_compile) tutorial.

Reviewer: given that the compilation instructions are in each of the sub-samples, you should remove the commands from here.
